WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -0,0 +1,274 @@
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Blockchain.Keys;
+using WalletWasabi.Crypto.Randomness;
+using WalletWasabi.Crypto.ZeroKnowledge;
+using WalletWasabi.Logging;
+using WalletWasabi.WabiSabi.Backend.PostRequests;
+using WalletWasabi.WabiSabi.Backend.Rounds;
+using WalletWasabi.WabiSabi.Crypto;
+using WalletWasabi.WabiSabi.Models;
+using WalletWasabi.WabiSabi.Models.MultipartyTransaction;
+using WalletWasabi.Wallets;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	public class CoinJoinClient : BackgroundService, IDisposable
+	{
+		private bool _disposedValue;
+
+		public CoinJoinClient(
+			uint256 roundId,
+			IWabiSabiApiRequestHandler arenaRequestHandler,
+			IEnumerable<Coin> coins,
+			Kitchen kitchen,
+			KeyManager keymanager)
+		{
+			RoundId = roundId;
+			ArenaRequestHandler = arenaRequestHandler;
+			Kitchen = kitchen;
+			Keymanager = keymanager;
+			SecureRandom = new SecureRandom();
+			Coins = coins;
+		}
+
+		private uint256 RoundId { get; }
+		private RoundState RoundState { get; set; }
+		private ZeroCredentialPool ZeroAmountCredentialPool { get; } = new();
+		private ZeroCredentialPool ZeroVsizeCredentialPool { get; } = new();
+		private SecureRandom SecureRandom { get; }
+		private CancellationTokenSource DisposeCts { get; } = new();
+		private IEnumerable<Coin> Coins { get; set; }
+		private Random Random { get; } = new();
+		public IWabiSabiApiRequestHandler ArenaRequestHandler { get; }
+		public Kitchen Kitchen { get; }
+		public KeyManager Keymanager { get; }
+
+		protected override async Task ExecuteAsync(CancellationToken cancellationToken)
+		{
+			try
+			{
+				await RefreshRoundAsync(cancellationToken).ConfigureAwait(false);
+				var constructionState = RoundState.Assert<ConstructionState>();
+
+				// Calculate outputs values
+				var outputValues = DecomposeAmounts();
+				var outputs = outputValues.Zip(Keymanager.GetKeys(), (amount, hdPubKey) => new TxOut(amount, hdPubKey.P2wpkhScript));
```

Comment:
This is not generating, keys but returns the existing ones.

Reply:
done.